test/quicksort.jl

```diff
@@ -0,0 +1,126 @@
+        sums_match &= post_sums[block] == count(x -> x >= pivot, temp)
+    end
+
+    @test sort_match
+    @test sums_match
+end
+
+full_size = quicksort_block_size(Float32)
+half_size = full_size ÷ 2
+test_batch_partition(Float32, 10000, half_size, 0, 10000, 0)
+test_batch_partition(Float32, 10000, half_size, 2000, 6000, 0)
+test_batch_partition(Float32, 10000, half_size, 2000, 6000, 1)
+test_batch_partition(Float32, 10000000, half_size, 0, 10000000, 0)
+test_batch_partition(Float32, 10000000, half_size, 5000, 500000, 0)
+test_batch_partition(Float32, 10000, full_size, 0, 10000, 0)
+test_batch_partition(Float32, 10000, full_size, 2000, 6000, 0)
+test_batch_partition(Float32, 10000, full_size, 2000, 6000, 1)
+test_batch_partition(Float32, 10000000, full_size, 0, 10000000, 0)
+test_batch_partition(Float32, 10000000, full_size, 5000, 500000, 0)
+
+@test seek_last_less_than([1, 2, 2, 3, 4, 1, 2, 2, 3, 4], 3, 0, 5, true) == 3
+@test seek_last_less_than([1, 2, 2, 3, 4, 1, 2, 2, 3, 4], 3, 5, 10, true) == 8
+
+function test_consolidate_kernel(vals, pivot, my_floor, L, N_t, N_b, b_sums, dest)
+    i = threadIdx().x
+    p = consolidate_batch_partition(vals, pivot, my_floor, L, N_t, N_b, i, b_sums)
+    if i == 1
+        dest[1] = p
+    end
+    return nothing
+end
+
+function test_consolidate_partition(T, N, block_dim, lo, hi, seed)
+    # assuming partition_batches works, we can validate consolidate by
+    # checking that together they partition a large domain
+    my_range = lo + 1 : hi
+    Random.seed!(seed)
+    original = rand(T, N)
+    A = CuArray(original)
+    block_N = ceil_div(hi - lo, block_dim)
+    pivot = rand(original[my_range])
+    sums = CuArray(zeros(Int32, block_N))
+
+    @cuda blocks=block_N threads=block_dim partition_batches_kernel(A, pivot, sums, lo, hi - lo, Val(block_dim), Val(true))
+    synchronize()
+    dest = CuArray(zeros(Int32, 1))
+    @cuda threads=block_dim test_consolidate_kernel(A, pivot, lo, hi - lo, block_dim, block_N, sums, dest)
+    synchronize()
+    partition = Array(dest)[1]
+    temp = original[my_range]
+    post_sort = Array(A)
+    #= consolidation is a highly unstable sort (again, by pivot comparison as
+    the key) so we compare by counting each element =#
+    @test counter(post_sort[lo + 1 : partition]) == counter(filter(x -> x < pivot, temp))
+    @test counter(post_sort[partition + 1 : hi]) == counter(filter(x -> x >= pivot, temp))
+end
+
+
+full_size = quicksort_block_size(Float32)
+test_consolidate_partition(Float32, 10000, full_size ÷ 4, 0, 10000, 0)
+test_consolidate_partition(Float32, 10000, full_size ÷ 2, 0, 10000, 0)
+test_consolidate_partition(Float32, 10000, full_size, 0, 10000, 0)
+
+full_size = quicksort_block_size(Int64)
+test_consolidate_partition(Int64, 10000, full_size ÷ 4, 0, 10000, 0)
+test_consolidate_partition(Int64, 10000, full_size ÷ 2, 0, 10000, 0)
+test_consolidate_partition(Int64, 10000, full_size, 0, 10000, 0)
```

Comment:
Maybe put all these tests in a `@testset "quicksort"` and also do some generic testing of `Base.sort` and `Base.sort!`, which is what most users will end up using.